### PR TITLE
Fixed form submission to be more resilient with abnormal inputs and transformValue properties

### PR DESCRIPTION
**Pull request by cjreimer**
As per issue #2162, when a form is submitted with a `DateField `or a `DateTimeLocalField `that is empty, it will normally cause a validation error.  However, if the validation is set to not required, it will cause an exception.

The code has been changed to return a `null `in this case.  If being sent to the API side, this may put a null in the database, which should be fine for a non-required datefield.

Testing for this case also added.

### CODE REVIEW

**Comment by Tobbe** [comment]
Another fix Curtis?! This is great! Thanks 🙂 

Do we need similar fixes for the other coercions? 

JSON.parse("") -> Uncaught SyntaxError
parseInt("", 10) -> NaN
parseFloat("") -> NaN

**Comment by cjreimer** [comment]
@Tobbe .... great points on related issues!  OK, let's do this!  

Really these issues come up because of inappropriate validation, and thus it is debatable what we should return.  
I'm proposing the following, and I've got a basic version of the code working for the following, but I'm open to suggestions as to how we handle these cases before I finalize and commit new code:

**JSON.parse("") -> Uncaught SyntaxError**

This needs to be handled at the validation stage.  So, if someone sets a Json coercion via the transformValue I propose we automatically set a corresponding validation function (unless they override).  Thus, we have two cases:
- Case 1:  Someone sets `transformValue="Json"` but no specific custom validation
    We handle this case via an internal JSON validation function.  The coercion function should only get valid JSON

- Case 2 (rare): Someone sets `transformValue="Json"`and a specific custom validation that lets through bad JSON
    We catch the failed coercion, console.warn a message and gracefully return an `undefined`.  I'm proposing an `undefined `as graphql and prims handle this gracefully, with prisma not making any database update on the field in an update.
  
**parseInt("", 10) -> NaN**

We normally should only get this case if someone turns off the "required" validation.  I suggest we return an "`undefined`" as prisma will handle this gracefully and not update the value.  We could force a 0, but I hate creating data that is not represented.

**parseFloat("") -> NaN**
Propose same response as parseInt

**Additional Check** 
Right now, if someone mistypes a transformValue input, say:
` transformValue="json"`

rather than
` transformValue="Json"`

The code handles it gracefully, but there are no warnings ... just no coercion happens.  This could easily take someone a while to figure out when debugging. How about doing a console.warn if someone tries to set a transformValue that is not valid.  It could be helpful to a developer.  What are your thoughts on using console.warn on the front end for these types of things?

**Comment by Tobbe** [comment]
I like your ideas Curtis. If we catch bad json and return `undefined`, should we do the same for parseInt/parseFloat? I.e. should we let 'five' be `NaN`, or should we change it to `undefined`?

I agree typing `json` with a lowercase j could be a difficult bug to spot. Some kind of warning, in development only, seems like a nice help.

@cannikin I'd like your input on Curtis' proposal as well. Personally I'm all for it. What do you say?

**Comment by cjreimer** [comment]
@Tobbe I agree for the parseInt/parseFloat, anytime we get NaN we should return `undefined`.

I also like your idea of the warnings being development only.  We can warn the developer if we run into these cases, as really the form and validation should be set up properly so we don't return `undefined `to the api long term.

File: packages/forms/src/coercion.tsx [hunk 1]
```diff
@@ -21,12 +21,48 @@ export const CoercionContextProvider: React.FC = ({ children }) => {
   )
 }
 
+const coercionWarn = (type: string, value: string) => {
+  if (process.env.NODE_ENV !== 'production') {
```
Comment (Tobbe) [inline packages/forms/src/coercion.tsx:25]: This has bitten us before. Please change to check if it's 'development' or 'testing'

File: packages/forms/src/coercion.tsx [hunk 2]
```diff
@@ -66,6 +102,11 @@ export const useCoercion = () => {
       } else {
         if (transformValue) {
           coercionFunction = COERCION_FUNCTIONS[transformValue]
+          if (!coercionFunction && process.env.NODE_ENV !== 'production') {
```
Comment (Tobbe) [inline packages/forms/src/coercion.tsx:105]: Same as above, dev or test

**Comment by cjreimer** [comment]
@Tobbe I believe I'm ready for this to be reviewed.  I did make a change from the original proposal to return `undefined `from a `DateTime `or `DateTimeLocalField`.  Thanks for your help!

I'll start on the docs update.

File: packages/forms/src/coercion.tsx (hunk 1 above)
Comment (cjreimer) [reply packages/forms/src/coercion.tsx:25]: Man you are fast!  Do you ever sleep?
I'll get on it!

File: packages/forms/src/coercion.tsx (hunk 2 above)
Comment (cjreimer) [reply packages/forms/src/coercion.tsx]: @Tobbe Got it, fixed.  Thanks.

File: packages/forms/src/index.tsx [hunk 3]
```diff
@@ -246,13 +258,18 @@ const TextAreaField = forwardRef<
   }, [setCoercion, props.name, props.transformValue, props.dataType])
 
   const tagProps = inputTagProps(props)
+  // implements JSON validation if a transformValue of 'Json' is set
+  const validation = props.validation ? props.validation : { required: false }
+  if (!validation.validate && props.transformValue == 'Json') {
```
Comment (Tobbe) [inline packages/forms/src/index.tsx:263]:
```suggestion
  if (!validation.validate && props.transformValue === 'Json') {
```

File: packages/forms/src/index.tsx [hunk 4]
```diff
@@ -315,7 +332,7 @@ export const CheckboxField = forwardRef<
   const type = 'checkbox'
 
   React.useEffect(() => {
-    if (process.env.NODE_ENV !== 'production' && props.dataType !== undefined) {
+    if (props.dataType !== undefined) {
```
Comment (Tobbe) [inline packages/forms/src/index.tsx:335]: How come you removed the NODE_ENV check entirely here?

File: packages/forms/src/index.tsx (hunk 3 above)
Comment (cjreimer) [reply packages/forms/src/index.tsx:263]: Sure

File: packages/forms/src/index.tsx (hunk 4 above)
Comment (cjreimer) [reply packages/forms/src/index.tsx:335]: Not sure ... my error.  Thanks for catching!

**Comment by cjreimer** [comment]
@Tobbe, do you mind giving me a pointer here?  I see the Cypress test failed.  From what I can see, I'm a little surprised my section of code caused the corresponding section of the Cypress tests to fail, but I'm open to chasing it down.  I'm not sure how to run the Cypress tests locally on Windows.  Contributing.md says:

`> Windows Not Supported: The command for this is written in bash and will not work on Windows.`

Any advice you can give on how to start narrowing this down?

Thanks!

**Comment by Tobbe** [comment]
> @Tobbe, do you mind giving me a pointer here? I see the Cypress test failed.

It sometimes does that on Windows :(  Just re-run it and it usually passes.

**Review by Tobbe — APPROVED** [review]

**Comment by thedavidprice** [comment]
Thanks for this @cjreimer 🚀

> I'll start on the docs update.

^^ is there a PR or Issue started for this? Let me know as I'd like to track it for the next release.

**Comment by Tobbe** [comment]
@thedavidprice https://github.com/redwoodjs/redwoodjs.com/pull/660

**Pull request merged**